### DISCUSSION TITLE
LFX Mentorship 2023 03-Sep-Nov Challenge - for #2689

### DISCUSSION THREAD

**Discussion by juntao** [WasmEdge/WasmEdge#2703] [Show and tell]
For those who want to apply for LFX mentorship for #2689, this is a pretest for the application.

1. Complete pre-test #2692 to demonstrate that you can write and build a basic Rust program.
2. Compile ffmpeg to a shared library (.so file for Linux)
3. Create a WasmEdge plugin in any language (C/C++/Rust) for one ffmpeg function of your choice.
4. Create a demo Rust app that can call the ffmpeg function from WasmEdge.

**Comment by Hrushi20** [comment]
Ffmpeg is split into various module (libavcodec, libavfilter etc). Can I pick up any module and implement its functionality for the test ? Is the goal of lfx mentorship to write functionalities for all modules so that they can be used from Rust and wasmedge?

**Comment by Hrushi20** [comment]
Hey @juntao, I've completed the pretest

- [x] Compile ffmpeg to a shared library
- [x] Created [WasmEdge Plugin](https://github.com/Hrushi20/WasmEdge-FFMPEG#readme) for Ffmpeg using C++
- [x] Created [Demo Rust App](https://github.com/Hrushi20/ffmpeg-rust#readme) to call the ffmpeg function from WasmEdge.

The Demo Rust App contain images of the result. I thank @hydai for helping me debug and fix config issues. 
It was a fun learning 😄

**Reply by juntao** [reply]
Thank you! @Hrushi20

**Comment by Hrushi20** [comment]
Hey @juntao,
I've updated the test with an example. The Rust code uses the [libavcodec/avcodec.h](https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/avcodec.h) C API via the WasmEdgeFfmpeg Plugin. It reads a video file and generates grayscale image frames.

**Comment by 1awesomeJ** [comment]
Good morning @juntao 
Please what are the deliverables for this pre test.
It looks to me like the 3 tasks given are things that will be done on our local machines.

I'm currently building ffmpeg( to shared library)

1. The WasmEdge plugin we're to build, will it be just for our local WasmEdge instance, or we have to push it upstream?

2. The Rust app that will use this plug in, how do we submit/demo it?

Thank you sir.

**Reply by juntao** [reply]
Hi @1awesomeJ 

You can create a new repo that contains the plugin. We will build the plugin and then install it with the latest WasmEdge.

Same for the Rust demo. You can submit code via your own GitHub repo. 

Cheers
Michael

**Reply by 1awesomeJ** [reply]
That's awesome.
Thank you sir.

**Comment by 1awesomeJ** [comment]
@juntao,
I finally got a "add" function I wrote in C to be successfully called
within a rust program!
The chance of being mentored by Dr Yuan is too amazing to lose!

So far I see I can compile my C code to web assembly using emcc, then call the ```run_func_from_file()``` method on a ```wasmedge_sdk::VmBuilder``` object passing in the path to my wasm file, the name of the C function and its parameters.

Issues:
a. I can only implement functions that take i32 as params and return i32 for now
.I need to learn how to implement C functions that take string params and custom type params.

b. I need to fully understand and implement the shared library approach of building wasmedge plugins.

Questions:
1. Do both C files compiled to wasm and those implemented as Host Functions and built into a shared library qualify as wasmedge plugins?

2. Can the plugin required for this pretest be built as a wasm?

Sorry I was verbose!

**Comment by 1awesomeJ** [comment]
@juntao 
Please I need a resource on how to use a loadble (".so") plugin in a rust application.

So far, I've only found these in the docs:
1. How to build a ".so" loadable plugin with C API and cmake. I couldn't find anything yet on how to use such plugin in an external application.
2. How to use external functions in a rust applications using the ```run_func_from_file()``` function from ```VmBuilder``` struct of the ```wasmedge_sdk``` crate. -I only saw it used with a wasm file.

**Comment by 1awesomeJ** [comment]
https://github.com/1awesomeJ/WasmEdge_ffmpeg_Plugin

@juntao, The link above contains my work on this pre test. It has my current work on the plugin. It'll also have my work on the rust app that uses the plugin when it's ready.

My level of progress will be seen in the repository whenever you have to do your review.

Thank you sir.

**Comment by macroexpansion** [comment]
Hi @juntao 
These are my pre-tests for #2689 mentorship,

1. WasmEdge with FFMPEG Plugin enabled written in C++ and compile to a shared library.
https://github.com/blueyellowpink/wasmedge-ffmpeg

2. WasmEdge FFMPEG SDK and a simple program converts `mp4` files to `ts` files written in Rust. It calls the WasmEdge FFMPEG Plugin shared library.
https://github.com/blueyellowpink/wasmedge-ffmpeg-sdk

Best regards,
Andrew

**Reply by juntao** [reply]
These are very cool. Thanks!

**Comment by alabulei1** [comment]
Hi @blueyellowpink, @Hrushi20, and @1awesomeJ , 

Thank you for completing the pretest. If you have not yet submitted your application for this project, please do so before 5 PM PDT on August 25th. 

Apply here: https://mentorship.lfx.linuxfoundation.org/project/41a8bf79-a903-4c03-afb7-256fd373c0b0

**Comment by xevor11** [comment]
Done with https://github.com/Vtewari2311/wasmedge-ffmpeg
